sklearn_numba_dpex/common/topk.py

```diff
@@ -782,253 +1000,361 @@ def _make_check_radix_histogram_kernel(radix_size, dtype):
+
+    @dpex.kernel
+    # fmt: off
+    def gather_topk(
+        array_in,                           # IN READONLY    (n_rows, n_cols)
+        threshold,                          # IN             (n_rows,)
+        n_threshold_occurences_in_topk,     # IN             (n_rows,)
+        n_threshold_occurences_in_data,     # IN             (n_rows,)
+        index_buffer,                       # BUFFER         (n_rows,)
+        result,                             # OUT            (n_rows, k)
+    ):
+        # fmt: on
+        row_idx = dpex.get_global_id(zero_idx)
+        col_idx = dpex.get_global_id(one_idx)
+
+        n_threshold_occurences_in_topk_ = n_threshold_occurences_in_topk[row_idx]
+
+        threshold_ = threshold[row_idx]
+
+        # Different branches depending on the value of `n_threshold_occurences`, since
+        # some optimizations are possible depending on the value.
+        if n_threshold_occurences_in_data[row_idx] == n_threshold_occurences_in_topk_:
+            gather_topk_include_all_threshold_occurences(
+                row_idx,
+                col_idx,
+                threshold_,
+                array_in,
+                index_buffer,
+                # OUT
+                result
+            )
+        else:
+            gather_topk_generic(
+                row_idx,
+                col_idx,
+                threshold_,
+                n_threshold_occurences_in_topk_,
+                array_in,
+                index_buffer,
+                # OUT
+                result,
+            )
 
     # Both the number of occurences of the threshold value in the data and the number
     # in the top-k values are known. When those two numbers are equal, the kernel can
     # be written more efficient and much simpler, and the condition is not unusual.
     # Let's write a separate kernel for this special case.
-    if n_threshold_occurences_in_topk == n_threshold_occurences_in_data:
-
-        @dpex.kernel
-        # fmt: off
-        def gather_topk_include_all_threshold_occurences(
-            array_in,                          # IN READ-ONLY (n_items,)
-            threshold,                         # IN           (1,)
-            n_threshold_occurences,            # UNUSED BUFFER
-            index_buffer,                      # BUFFER       (1,)
-            result,                            # OUT          (k,)
-        ):
-            # fmt: on
-            item_idx = dpex.get_global_id(zero_idx)
-
-            if item_idx >= n_items:
-                return
-
-            if index_buffer[zero_idx] >= k:
-                return
+    @dpex.func
+    # fmt: off
+    def gather_topk_include_all_threshold_occurences(
+        row_idx,                           # PARAM
+        col_idx,                           # PARAM
+        threshold,                         # PARAM
+        array_in,                          # IN READ-ONLY (n_items,)
+        index_buffer,                      # BUFFER       (n_rows,)
```

Comment:
Similar changes to do in other places if you agree with this suggested change.

Reply:
The naming of `index_buffer` is discussed in another suggestion.

Reply:
I dislike INOUT here because having OUT suggests that the values stored in the array will be used later outside the kernel call, which is not the case here, this array really just is a buffer used for synchronizing threads globally. Maybe TMP ?